feat(faucet): api endpoint to return distribution

iancoleman opened the pull request:
The server sends 100 tokens for the endpoint:
`http://<ip>/<bls_hex_pubkey>`

This adds a new (additional) endpoint to distribute to maid addresses:
`http://<ip>/distribution?address=<maid_address>&pkhex=<pkhex>`

Calling the new endpoint will return an existing distribution for that maidsafecoin address, or create a new distribution if needed.

This enables distributions to be fetched from the faucet, but there still needs to be integration into the cli which is future work to try to keep each PR relatively small and manageable.

Future faucet related work:

* update cli `safe wallet get-faucet` to be able to fetch distributions
* change the default amount issued by the faucet from 100 to something smaller (any suggestions?)
* potentially move away from `tiny_http` for the faucet server to something that has better security and rate limiting options

## Description

reviewpad:summary

joshuef commented on sn_node/src/bin/sn_node.rs:
```diff
@@ -310,9 +310,11 @@ async fn run_node(config: Config) -> Result<()> {
+
     let (node, mut event_stream) = loop {
-        match NodeApi::new(&config, bootstrap_retry_duration).await {
+        match NodeApi::new(&config, join_timeout).await {
```
Why have we a new variable that's twice the intended retry time here?

----

Previously:

if `bootstrap_retry_duration` was hit in the internal `timeout`, we errored. We then wait `bootstrap_retry_duration` again before trying.

Here:

If after `2 * bootstrap_retry_duration` we haven't  joined, we error (out of the loop 👍 ). And then wait `bootstrap_retry_duration` once more? Is that correct?

Is there a need for a new variable twice as large? (I appreciate naming can be improved here regardless). What's the reason for `2*` ?

oetyng replied:
It just seemed as if a varable used for one thing (bootstrap retry interval) had been used for another (join timeout).
As I've seen the timeout error now and then, I thought 30s could be used  and seemed like a valid timeout.
Better as a new const but didn't seem overly important considering the already existing reuse of the const for separate things.

oetyng replied:
Added separate const.

joshuef replied:
Makes sense. I think that's probably my actual confusion here, the initial confused naming.

`bootstrap_retry_duration` sounds like _over this time, we'll retry. That wasn't accurate before... it was twice as long from initial contact attempt to retry, and now it's three times as long.

Which feels off, but the source of this isn't this PR I think, so feel free to ignore this. I'm not sure I have any ideas of to improve the naming / use there (and as `bootstrap_retry_duration` isn't a config option atm... I think it doesn't really matter).

joshuef commented on sn_node/src/node/core/bootstrap/join.rs:
```diff
@@ -152,14 +162,22 @@ impl<'a> Join<'a> {
 
         loop {
-            let (response, sender) = self.receive_join_response().await?;
+            join_timeout = join_timeout
```
Can you add a comment just clarifying the intended function here. That we error out if we've exceeded the `join_timeout` 🙇

joshuef commented on sn_node/src/node/core/bootstrap/join.rs:
```diff
@@ -434,8 +452,21 @@ impl<'a> Join<'a> {
+    ) -> Result<(JoinResponse, Peer)> {
+        let mut timer = Instant::now();
+
+        while let Some(event) = tokio::time::timeout(join_timeout, self.incoming_msgs.recv())
```
same here please w/r/t comment

joshuef commented:
merged via #1292

iancoleman committed:
feat(faucet): api endpoint to return distribution

The server sends 100 tokens for the endpoint:
http://<ip>/<bls_hex_pubkey>
This adds a new (additional) endpoint to distribute to maid addresses:
http://<ip>/distribution?address=<address>&pkhex=<pkhex>
Calling the new endpoint will return an existing distribution for that
maidsafecoin address, or create a new distribution if needed.

joshuef commented:
> change the default amount issued by the faucet from 100 to something smaller (any suggestions?)

I think we should probably look to distribute 1SNT only and that from a wallet that only has maybe 30 to start. (arbitrary numbers!)

I'm thinking: 

- we do not want an endless faucet. Indeed we only really want some for basic in house testing.
- we can top up a faucet if desired, so : keep the amount low (1 seems sensible atm?), and the initial dist to the hot-faucet-wallet big enough, but not wild.
- this will force folk to use and test out the OMNI conversion.


Thoughts?

iancoleman commented:
> I think we should probably look to distribute 1SNT only

I think it should be much lower than that, more like 0.01 SNT

In recent testnets upload cost has been about 0.002 SNT per GB

> from a wallet that only has maybe 30 to start

This sounds about right. Probably could even start with less. Has the faucet run dry before? Have we got a code path for this (I guess we do of some sort but it's 'accidental' like a failed transfer)?
